assets/js/amp-block-validation.js

```diff
@@ -0,0 +1,140 @@
+/**
+ * Validates blocks for AMP compatibility.
+ *
+ * This uses the REST API response from saving a page to find validation errors.
+ * If one exists for a block, it display it inline with a Notice component.
+ */
+
+/* exported ampBlockValidation */
+var ampBlockValidation = ( function() {
+	'use strict';
+
+	var module = {
+		/**
+		 * Holds data.
+		 */
+		data: {},
+
+		/**
+		 * The blocks with validation errors.
+		 */
+		blocksWithErrors: {},
+
+		/**
+		 * Boot module.
+		 *
+		 * @param {Object} data Object data.
+		 * @return {void}
+		 */
+		boot: function( data ) {
+			module.data = data;
+			wp.data.subscribe( function() {
+				module.blocksWithErrors = module.getBlocksWithErrors();
+			} );
+			wp.hooks.addFilter(
+				'blocks.BlockEdit',
+				'amp/add-notice',
+				module.conditionallyAddNotice
+			);
+		},
+
+		/**
+		 * Wraps the edit() method of a block, and conditionally adds a Notice.
+		 *
+		 * @param {Function} OriginalBlockEdit The original edit() method of the block.
+		 * @returns {Function} The edit() method, conditionally wrapped in a notice for AMP validation error(s).
+		 */
+		conditionallyAddNotice: function( OriginalBlockEdit ) {
+			return function( props ) {
+				var errors = module.getBlockValidationErrors( props ),
+					result = [ wp.element.createElement( OriginalBlockEdit, props ) ];
+
+				if ( errors.length > 0 ) {
+					result.push( wp.element.createElement(
+						wp.components.Notice,
+						{
+							status: 'warning',
+							content: module.data.i18n.notice.replace( '%s', props.name ) + ' ' + module.blocksWithErrors[ props.name ][0].code,
+							isDismissible: false
+						}
+					) );
+				}
+				return result;
+			};
+		},
+
+		/**
+		 * Gets the block types with errors.
+		 *
+		 * Iterates through the 'amp_validation_errors' from the REST API response.
+		 * This returns an object, with block types as the keys, and error arrays as the values.
+		 * The block's overridden edit() method can then get the errors for its block type.
+		 *
+		 * @returns {Object} The blocks with errors.
+		 */
+		getBlocksWithErrors: function() {
+			var currentPost      = wp.data.select( 'core/editor' ).getCurrentPost(),
+				blocksWithErrors = {};
+			if ( ! currentPost.hasOwnProperty( module.data.errorKey ) || ! Array.isArray( currentPost[ module.data.errorKey ] ) ) {
+				return blocksWithErrors;
+			}
+
+			currentPost[ module.data.errorKey ].forEach( function( validationError ) {
+				if ( validationError.hasOwnProperty( 'sources' ) ) {
+					validationError.sources.forEach( function( source ) {
+						if ( source.hasOwnProperty( 'block_name' ) ) {
+							if ( blocksWithErrors.hasOwnProperty( source.block_name ) ) {
+								blocksWithErrors[ source.block_name ].push( validationError ); // jscs:ignore requireCamelCaseOrUpperCaseIdentifiers
+							} else {
+								blocksWithErrors[ source.block_name ] = [ validationError ]; // jscs:ignore requireCamelCaseOrUpperCaseIdentifiers
```

Comment:
I think it would be very useful to capture the `block_attrs` as well here. The `block_name` and `block_attrs` together are what could help to narrow down the validation errors for a given block in `getBlockValidationErrors`.

Reply:
Thanks, @westonruter. That's a good idea. [These commits](https://github.com/Automattic/amp-wp/compare/743cca49a132a147085125a274e23a0bb62c057c...b9aa16c39b1452b0eee17c76305bba48f5918887) add `block_attrs` to `blocksWithErrors`.